pkg/synthetictests/historicaldata/next_best_guess.go

```diff
@@ -0,0 +1,163 @@
+	if len(in.FromRelease) == 0 {
+		return platformidentification.JobType{}, false
+	}
+
+	fromReleaseMinor := getMinor(in.FromRelease)
+	toReleaseMajor := getMajor(in.Release)
+	toReleaseMinor := getMinor(in.Release)
+	// if we're already a minor upgrade, this doesn't apply
+	if fromReleaseMinor == (toReleaseMinor - 1) {
+		return platformidentification.JobType{}, false
+	}
+
+	ret := platformidentification.CloneJobType(in)
+	ret.FromRelease = fmt.Sprintf("%d.%d", toReleaseMajor, toReleaseMinor-1)
+	return ret, true
+}
+
+// MicroReleaseUpgrade if we don't have data for the current fromRelease and it's a minor upgrade, perhaps we have data
+// for a minor upgrade.
+func MicroReleaseUpgrade(in platformidentification.JobType) (platformidentification.JobType, bool) {
+	if len(in.FromRelease) == 0 {
+		return platformidentification.JobType{}, false
+	}
+
+	fromReleaseMinor := getMinor(in.FromRelease)
+	toReleaseMajor := getMajor(in.Release)
+	toReleaseMinor := getMinor(in.Release)
+	// if we're already a micro upgrade, this doesn't apply
+	if fromReleaseMinor == toReleaseMinor {
+		return platformidentification.JobType{}, false
+	}
+
+	ret := platformidentification.CloneJobType(in)
+	ret.FromRelease = fmt.Sprintf("%d.%d", toReleaseMajor, toReleaseMinor)
+	return ret, true
+}
+
+// PreviousReleaseUpgrade if we don't have data for the current toRelease, perhaps we have data for the congruent test
+// on the prior release.
+func PreviousReleaseUpgrade(in platformidentification.JobType) (platformidentification.JobType, bool) {
+	if len(in.FromRelease) == 0 {
+		return platformidentification.JobType{}, false
+	}
+
+	fromReleaseMinor := getMinor(in.FromRelease)
+	toReleaseMajor := getMajor(in.Release)
+	toReleaseMinor := getMinor(in.Release)
+
+	ret := platformidentification.CloneJobType(in)
+	ret.Release = fmt.Sprintf("%d.%d", toReleaseMajor, toReleaseMinor-1)
+	ret.FromRelease = fmt.Sprintf("%d.%d", toReleaseMajor, fromReleaseMinor-1)
+	return ret, true
+}
+
+func getMajor(in string) int {
+	major, err := strconv.ParseInt(strings.Split(in, ".")[0], 10, 32)
+	if err != nil {
+		panic(err)
+	}
+	return int(major)
+}
+
+func getMinor(in string) int {
+	minor, err := strconv.ParseInt(strings.Split(in, ".")[1], 10, 32)
+	if err != nil {
+		panic(err)
+	}
+	return int(minor)
+}
+
+// OnOVN maybe we have data on OVN
+func OnOVN(in platformidentification.JobType) (platformidentification.JobType, bool) {
+	if in.Network == "ovn" {
+		return platformidentification.JobType{}, false
+	}
+
+	ret := platformidentification.CloneJobType(in)
+	ret.Network = "ovn"
+	return ret, true
+}
+
+// OnSDN maybe we have data on SDN
+func OnSDN(in platformidentification.JobType) (platformidentification.JobType, bool) {
+	if in.Network == "sdn" {
+		return platformidentification.JobType{}, false
+	}
+
+	ret := platformidentification.CloneJobType(in)
+	ret.Network = "sdn"
+	return ret, true
+}
+
+func combine(nextBestKeys ...NextBestKey) NextBestKey {
```

Comment:
I can't follow what you're trying to do with combine, it's used above with combos of matchers we already processed, which if they returned true, would have been used. Then you "combine" them, which looks to be logic for "if any of the combined guessers returns false, return false, else return the last one that was true", but this would never run because the guessers were already processed individually and returned true?

I'm likely missing something but I've stared at it long enough to know it really needs some godoc, here on this func and probably up above in the list as well.

Reply:
combine runs the first and uses the output for the next.  I'll doc it.  That allows you to have methods for same update last release and for minor instead of micro, and combined them to have do you have a minor upgrade from last release.